### ISSUE TITLE
Dynamic Background Video Link for Sections

### ISSUE THREAD

**Issue by JordanKDesigns** [elementor/elementor#6116]
## Prerequisites

- [x] I have searched for similar features requests in both open and closed tickets and cannot find a duplicate.
- [x] The feature is still missing in the latest stable version of Elementor ( Elementor Pro. )

It would be great if we could get a dynamic option for the background video link AND fallback image for the section elements. Extending this request, I think a dynamic option for almost everything would be great.

![elementor](https://user-images.githubusercontent.com/44751665/48280002-25824b00-e407-11e8-81d5-ebb0e54a9c90.png)

**Comment by silvioghedin** [comment]
+1

**Comment by gwashbrook** [comment]
+1

**Comment by laurentvergnaud** [comment]
+1

**Comment by laurentvergnaud** [comment]
+1

**Comment by alriksson** [comment]
+1 and support local hosted/ Cloudflare stream video for performance.

**Comment by harryneufeld** [comment]
+1

**Comment by Joe-Bloggs** [comment]
OMG +100

**Comment by seankbennett** [comment]
Yes please!

**Comment by artifex-media** [comment]
+1

**Comment by kldany** [comment]
i have requested the same i think it will be great in that option will be added

**Comment by kldany** [comment]
i found this but not sure it does the needs 
https://codecanyon.net/item/elementor-background-image-video-slider/21598997

**Comment by kldany** [comment]
By the way, you can use this solution
1- Download the video you need from youtube 
2- Give the video a (bg-video) Name or any name
3- uploaded it to the media folder, "I recommend creating videos folder"
4- copy the link and paste it in Elementor section video link holder
 its works and you can rename any downloaded video and replace the file
Check this sample: https://thecarsdealer.com

Faris

**Comment by kostas33** [comment]
+1000 PLEASE... this is really necessary guys

**Comment by fvenuto** [comment]
+1!!! 
I just need exactly this for single post template.
Does someone maybe have a workaround in the meanwhile?
Thanks.

**Comment by davidchc** [comment]
An alternative I found was to enable shortcode for video link (background_video_link). In this way it is possible to work defining the video using the shortcode, making this value dynamic. Example: [acf field = "introduction_video"]. Add this code to your theme's functions.php file.

```php
<?php
add_action( 'init', function(){
    add_action( 'elementor/frontend/section/before_render', function ( \Elementor\Element_Base $element ) {
        if ( ! $element->get_settings( 'background_video_link' ) ) {
            return;
        }
        $element->set_settings( 'background_video_link',  do_shortcode($element->get_settings( 'background_video_link' ) ) );
    } );
});```

**Comment by jakebradley2001** [comment]
> An alternative I found was to enable shortcode for video link (background_video_link). In this way it is possible to work defining the video using the shortcode, making this value dynamic. Example: [acf field = "introduction_video"]. Add this code to your theme's functions.php file.
> 
> ```
> <?php
> add_action( 'init', function(){
>     add_action( 'elementor/frontend/section/before_render', function ( \Elementor\Element_Base $element ) {
>         if ( ! $element->get_settings( 'background_video_link' ) ) {
>             return;
>         }
>         $element->set_settings( 'background_video_link',  do_shortcode($element->get_settings( 'background_video_link' ) ) );
>     } );
> });
> ```

Hi David, this looks like a great solution for what I'm looking for, but I'm confused on where you place the shortcode after putting the code in your functions.php. Mainly I'm looking for a solution that if someone choose an image in their ACF field, it would do a image background and if they choose to do a video, it would do video, suggestions?

**Comment by davidchc** [comment]
@jakebradley2001 , the code snippet is to be added to the functions.php file, to enable it to be possible to interpret the shortcode added in that field of the video.

About changing colors, I know it is possible to do this using the PRO version, where you define assignments, and you can choose a custom field, being able to use the ACF. And in the advanced settings of the attributes, you can define: style | background:.
In this way the elementor transforms it into style = "background: Color defined Shortcode"

**Comment by chloeford0** [comment]
+1

**Comment by kostas33** [comment]
@davidchc Hi there, first of all, thank you for helping with this. I am trying to implement it in my site. I added your code to the functions.php, i have an ACF field (type= url and slug= background_video) and i added this shortcode ini the elementor [acf field = "background_video"] . What am i doing wrong here?? i get no video... sorry.. i am a real noob at this.. thanks in advance!

**Comment by kostas33** [comment]
@davidchc the problem is that the video is in categories, category archive page,  not posts... :(

**Comment by davidchc** [comment]
@kostas33 , I wouldn't be able to tell you how shortcode acf behaves in an archive, for example. My suggestion is to actually test if it will render a shortcode directly.

**Comment by kostas33** [comment]
@davidchc ok, i will try it, thank you very much

**Comment by kostas33** [comment]
Just for those who still need it, i managed to get it working with method below and the code from @davidchc (thanks again mate). 

So, the steps are

1. I created an acf field for the taxonomy (background_video)
2. In the functions.php i added the code from davidchc with an addition to get the field
3. I created a shortcode with the field 
4. I added the shortcode to the elementor's background video field ( [cat-video] )

The final code is
`add_action( 'init', function(){
    add_action( 'elementor/frontend/section/before_render', function ( \Elementor\Element_Base $element ) {
        if ( ! $element->get_settings( 'background_video_link' ) ) {
            return;
        }
        $element->set_settings( 'background_video_link',  do_shortcode($element->get_settings( 'background_video_link' ) ) );
    } );
});


function acf_cat_video( $atts ){
	$term = get_queried_object();
	$video = get_field('background_video', $term);
	return $video;
}
add_shortcode( 'cat-video', 'acf_cat_video' );`

I hope it works for you too. Thank you all.

**Comment by SorryGoneSurfing** [comment]
Also struggling with this. 

I've set up a **custom field** 'genre_video' assigned to a **custom post type** (genre) set up with the CPT UI plugin. I've added the code @davidchc wrote in to the function file. 

When I call [acf field = "genre_video"]  in a text field it returns the correct value, but when I add this shortcode in the background video setting, I get nothing. Not sure what step I'm missing.. 


add_action('init', function () {
	add_action('elementor/frontend/section/before_render', function (\Elementor\Element_Base $element) {
		if (!$element->get_settings('genre_video')) {
			return;
		}
		$element->set_settings('genre_video',  do_shortcode($element->get_settings('genre_video')));
	});
});
add_shortcode('genre-video', 'acf_genre_video');

**Comment by adverman** [comment]
This **works** for me, without shortcode.
I stored the video-url in a simple text field and I follow the @davidchc code.
I used the `wp` hook because on `init` the WP Object has not yet been instantiated.

```
add_action( 'wp', function(){
  
   add_action( 'elementor/frontend/section/before_render', function ($element) {
      
      global $post;
      $video = get_post_meta($post->ID,'video-url',true);
      
      $element->set_settings( 'background_video_link',  $video);

   });

});
```

**Comment by bentalgad** [comment]
+1  Why not just add dynamic values every where? why should we ask for dynamic values for each field separately? just add them all! this 2020...

**Comment by adverman** [comment]
The issue seems to be correct with Elementor 2.9.9

**Comment by bentalgad** [comment]
> The issue seems to be correct with Elementor 2.9.9

I don't see a dynamic link option on 2.9.9 can you add a screenshot @adverman?

**Comment by adverman** [comment]
Dynamic link option has always been there, it just didn't work. Instead it now works.
<img width="317" alt="Schermata 2020-05-25 alle 15 59 19" src="https://user-images.githubusercontent.com/8201505/82819532-ced14a80-9ea0-11ea-8024-fa87975687cc.png">

**Comment by bentalgad** [comment]
Can you screenshot the whole tab, because i don't think we are talking about the same thing.

Is it the section style tab?

**Comment by adverman** [comment]
Yes!
<img width="345" alt="Schermata 2020-05-25 alle 16 35 11" src="https://user-images.githubusercontent.com/8201505/82822230-b9125400-9ea5-11ea-881c-833327b1f05f.png">

**Comment by bentalgad** [comment]
Thanks! i'm not sure why but i don't see it...


‫בתאריך יום ב׳, 25 במאי 2020 ב-17:35 מאת ‪adverman‬‏ <‪
notifications@github.com‬‏>:‬

**Comment by Joe-Bloggs** [comment]
This is really weird, I don't see it either!

@adverman do you possibly have any additional Elementor extensions installed?

**Comment by adverman** [comment]
Ehm... It needs the PRO version of Elementor

**Comment by Joe-Bloggs** [comment]
> Ehm... It needs the PRO version of Elementor

I do have a PRO version, still can't see it ..

**Comment by barchuuu** [comment]
I don't see it either :(
I realy need that feature!

**Comment by twotenstudio** [comment]
Also having this issue. I have latest version of pro and don't have the dynamic link option. Is there an additional plugin? A hidden setting?

**Comment by Joe-Bloggs** [comment]
So I contacted Elementor Pro's support about this and they have confirmed that currently, there is no such feature.

So again @adverman , do you possibly have any additional Elementor extensions, add-ons or plugins installed?

**Comment by adverman** [comment]
Only now did I realize that in fact it is so.
I believe functionality is introduced by this plugin https://www.dynamic.ooo/

**Comment by Joe-Bloggs** [comment]
> Only now did I realize that in fact it is so.
> I believe functionality is introduced by this plugin https://www.dynamic.ooo/

@adverman yes, I believe that's what it is! 

How is the plugin working out for you by the way? 
I've had this add-on installed before myself, but I found it quite heavy and too many bugs, so I got rid off it. That was about 6 months ago.

**Comment by adverman** [comment]
I like it a lot and it is constantly evolving ... together with croblock it is a must have.

**Comment by soulfulwater** [comment]
+1

**Comment by Pab89** [comment]
+1

**Comment by hagsonz** [comment]
must feature. by then you can use this snippet to check it the field contains shortcode. use acf\pods shortcode field to insert your video url.
`add_action('init', function () {
	add_action('elementor/frontend/section/before_render', function (\Elementor\Element_Base $element) {
		if (!$element->get_settings('background_video_link')) {
			return;
		}

	$shortcode_pattern = get_shortcode_regex();
        if ( preg_match('/'.$shortcode_pattern.'/s', $element->get_settings('background_video_link')) ) {
    		$element->set_settings('background_video_link',  do_shortcode($element->get_settings('background_video_link')));
        }

	});
});`

**Comment by hurtrez** [comment]
Hi
On behalf of a client, I have to create this page principle that he particularly appreciates :

https://www.molecularproducts.com/

Frozen vertical bands that come alive when the mouse passes over them, both by the video and by the movement that the latter generates.
Each mouse click will of course send to a specific page.
I thought Elementor made it easy to do, but now I'm blocking.

Thank you for your help

**DorShahar commented 3 hours ago**
@hurtrez
You can look at other plugins, like this one: https://codecanyon.net/item/unleash-a-wordpress-responsive-accordion-slider/12688077

Thanks

**Y write**
And thank you for this feedback
unfortunately the links appear to be dead. I watched the video that introduces the concept, it seems to match, great ... but what guarantee that it works ...? I will dig the thing
Thank you already for your responsiveness
I also note that the video has photo integrations and not videos ... too bad
but in any case, thank you for the approach
Someone might have a better idea ;-)

thank you for your help

**Comment by webdesign021pk** [comment]
> must feature. i tried 2 approaches, first is using a shortcode and the second is to replace elementor's video link control with a one that supports dynamic tags. for the shortcode approach, somehow Youtube videos didn't work as they didn't pass the regex expression used in `Embed::get_video_properties( $settings['background_video_link'] )` line at `section.php` file.
> anyways, i am posting both of the snippets, the first is recommended.
> 
> replacing the control approach:
> 
> ```
> // add dynamic tag to elementor background video link
> add_action( 'elementor/element/section/section_background/before_section_end', function( $element, $args ) {
> 	$element->remove_control('background_video_link');
> 
> 	$element->start_injection( [
> 		'type' => 'control',
> 		'at' => 'before',
> 		'of' => 'background_video_start',
> 	] );
> 
> 	// add a control
> 	$element->add_control(
> 		'background_video_link',
> 		[
> 			'label' => _x( 'Video Link', 'Background Control', 'elementor' ),
> 			'type' => \Elementor\Controls_Manager::TEXT,
> 			'description' => __( 'YouTube/Vimeo link, or link to video file (mp4 is recommended).', 'elementor' ),
> 			'label_block' => true,
> 			'default' => '',
> 			'dynamic' => [
> 					'active' => true,
> 					'categories' => [
> 							\Elementor\Modules\DynamicTags\Module::TEXT_CATEGORY,
> 							\Elementor\Modules\DynamicTags\Module::URL_CATEGORY,
> 					],
> 			],
> 			'render_type' => 'ui',
> 			'of_type' => 'video',
> 			'frontend_available' => true,
> 		]
> 	);
> 
> 	$element->end_injection();
> }, 10, 3 );
> ```
> 
> shortcode approach:
> 
> ```
> 
> add_action('init', function () {
> 	add_action('elementor/frontend/section/before_render', function (\Elementor\Element_Base $element) {
> 		if (!$element->get_settings('background_video_link')) {
> 			return;
> 		}
> 
> 		$shortcode_pattern = get_shortcode_regex();
>                 if ( preg_match('/'.$shortcode_pattern.'/s', $element->get_settings('background_video_link')) ) {
>     		        $element->set_settings('background_video_link',  do_shortcode($element->get_settings('background_video_link')));
>                 }
> 
> 	});
> });
> ```

Thanks a lot :) approach 1 worked great

**Comment by DorShahar** [comment]
Hey all,

This issue has been resolved in Elementor v3.2.0

Feel free to update!
